File: ObjCExceptionCatch/ObjCExceptionCatch.m
//
//  ObjCExceptionCatch.h
//  ObjCExceptionCatch
//
//  Copyright © 2018 Advantage FSE. All rights reserved.
//

#import "ObjCExceptionCatch.h"

@implementation ObjC

+ (BOOL)catchException:(void (^_Nonnull)(void))tryBlock error:(NSError **)error
{
    if (!tryBlock) {
        return YES;
    }

    @try {
        tryBlock();
        return YES;
    }
    @catch (NSException *exception) {
        if (error) {
            NSDictionary *exceptionUserInfo = exception.userInfo;
            NSMutableDictionary * userInfo = [exceptionUserInfo mutableCopy];
            userInfo[NSLocalizedDescriptionKey] = exception.reason;
            userInfo[NSUnderlyingErrorKey] = exception.name;
            *error = [[NSError alloc] initWithDomain:exception.name
                                                code:0
                                            userInfo:userInfo];
        }

        return NO;
    }
}

@end

File: ObjCExceptionCatch/ObjCExceptionCatch.swift
//
//  ObjCExceptionCatch.swift
//  ObjCExceptionCatch
//
//  Created by Christos Chryssochoidis on 07/03/2019.
//  Copyright © 2019 Advantage FSE. All rights reserved.
//

import Foundation

/**
    - param block: Needs to be @escaping otherwise we won't be able to pass throwing closures.
 */
public func objcTry<T>(block: @escaping () throws -> T) throws -> T {
    var result: T?
    var error: Error?
    
    let objcBlock: () -> () = {
        do {
            result = try block()
        } catch let anError{
            error = anError
        }
    }
    
    do {
        try ObjC.catchException(objcBlock)
    } catch let anError {
        error = anError
    }
    
    if let _result = result {
        return _result
    } else if let _error = error {
        throw _error
    } else {
        throw NSError(domain: "ObjCExceptionCatch", code: 0, userInfo: nil)
    }
}
